### PR TITLE
chore: use eslint 9 internally

### LINKED ISSUE #9082: Repo: Use ESLint v9 for our internal linting
### Suggestion

Now that #9002 is merged, our `v8` branch supports ESLint v9. Yay! But, #9002 didn't actually bump our _internal_ `devDependency`. https://github.com/typescript-eslint/typescript-eslint/pull/9002#discussion_r1591704965 mentions the three plugins whose lack of ESLint v9 support were blocking us.

But, now that https://eslint.org/blog/2024/05/eslint-compatibility-utilities exists, we should be able to move to ESLint v9 without losing our preferred linting. Which would be great. Accepting PRs to our `v8` branch to bring back the bump of our `"eslint"` `devDependency` from 8 to 9.

💖

### PR DESCRIPTION
**Pull request by abrahamguo**
## PR Checklist

- [x] Addresses an existing open issue: fixes #9082 
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview
Use eslint@9 internally.

### CODE REVIEW

**Comment by bradzacher** [comment]
We should delete this CI job
https://github.com/typescript-eslint/typescript-eslint/blob/87a0a484233ba76391be2d3a1588302051888485/.github/workflows/ci.yml#L137-L154

have we got an existing integration test to check for regressions on eslint v8?

**Comment by abrahamguo** [comment]
@bradzacher there are currently no integration or CI tests for eslint 8. I can add one if you want, but do you feel that's necessary since we ask for eslint 9 in both our peerDependencies and devDependencies?

Doesn't seem like anything similar was added for eslint 7 when we moved from eslint 7=>8 (#3737) 

Also, seems unrelated to this PR since this is about using eslint 9 _internally_

**Comment by bradzacher** [comment]
Looks like there's already an integration test for v8 - https://github.com/typescript-eslint/typescript-eslint/tree/main/packages/integration-tests/fixtures/eslint-v8
So nothing to add!

> Also, seems unrelated to this PR since this is about using eslint 9 internally

By moving to v9 internally our CI is no longer running against v8 - so we lose the test coverage asserting that we don't regress v8 compat.

But given we have the above integration test - we still have enough coverage.
It's not a super exhaustive test but it'll do (it's the same integration test coverage we maintained when we migrated to v8 internally)

**Comment by mattbrannon** [comment]
Just wanted to point out that one of your Netlify jobs has been running for nearly a week.
https://github.com/typescript-eslint/typescript-eslint/pull/9119/checks?check_run_id=25428243220

**Comment by abrahamguo** [comment]
@mattbrannon that's nothing compared to #8673, whose job has been running for 3 weeks 😉

It does seem like that specific job sometimes hangs indefinitely. Not sure why

File: eslint.config.mjs [hunk 1]
```diff
@@ -29,15 +30,20 @@ export default tseslint.config(
     plugins: {
       ['@typescript-eslint']: tseslint.plugin,
       ['@typescript-eslint/internal']: tseslintInternalPlugin,
-      ['deprecation']: deprecationPlugin,
+      // https://github.com/gund/eslint-plugin-deprecation/issues/85
+      // https://github.com/typescript-eslint/typescript-eslint/issues/8988
+      ['deprecation']: fixupPluginRules(deprecationPlugin),
```
Comment (abrahamguo) [inline eslint.config.mjs:35]: TODO: looks like this got fixed — try it out

File: eslint.config.mjs [hunk 2]
```diff
@@ -29,15 +30,20 @@ export default tseslint.config(
...
       ['@typescript-eslint/internal']: tseslintInternalPlugin,
-      ['deprecation']: deprecationPlugin,
+      // https://github.com/gund/eslint-plugin-deprecation/issues/78
+      // https://github.com/typescript-eslint/typescript-eslint/issues/8988
+      ['deprecation']: fixupPluginRules(deprecationPlugin),
       ['eslint-comments']: eslintCommentsPlugin,
       ['eslint-plugin']: eslintPluginPlugin,
-      ['import']: importPlugin,
+      // https://github.com/import-js/eslint-plugin-import/issues/2948
+      ['import']: fixupPluginRules(importPlugin),
```
Comment (bradzacher) [inline eslint.config.mjs:39]: we might want to consider switching to `eslint-plugin-import-x`?

File: eslint.config.mjs [hunk 3]
```diff
@@ -511,8 +517,8 @@ export default tseslint.config(
     files: ['packages/website/**/*.{ts,tsx,mts,cts,js,jsx}'],
     extends: [
       ...compat.config(jsxA11yPlugin.configs.recommended),
-      ...compat.config(reactPlugin.configs.recommended),
-      ...compat.config(reactHooksPlugin.configs.recommended),
+      ...fixupConfigRules(compat.config(reactPlugin.configs.recommended)),
+      ...fixupConfigRules(compat.config(reactHooksPlugin.configs.recommended)),
```
Comment (bradzacher) [inline eslint.config.mjs:521]: it's so weird that we need to fixup the plugin both here and in the plugin def.
seems like quite the flaw in the design of the compat tooling tbh

File: packages/website-eslint/build.ts [hunk 4]
```diff
@@ -31,9 +31,7 @@ function createResolve(
   join: string,
 ): esbuild.OnResolveResult {
   const resolvedPackage = requireResolved(targetPath + '/package.json');
-  return {
-    path: path.join(resolvedPackage, '../src/', join),
-  };
+  return { path: path.join(resolvedPackage, '../src/', join) };
```
Comment (JoshuaKGoldberg) [inline packages/website-eslint/build.ts:34]: [Style] This looks like an unrelated styling change. Could you please revert it and any others? It's harder to review PRs that have unnecessary whitespace changes - especially these larger, more intricate PRs.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -134,25 +134,6 @@ jobs:
         run: yarn stylelint
         working-directory: packages/website
 
-  eslint_v9_tests:
```
Comment (JoshuaKGoldberg) [inline .github/workflows/ci.yml:137]: [Testing] We'll still want to test against ESLint 8. I don't think anything would now do that? Unless I'm missing something, this should be switched to `eslint_v8_tests`, not removed.

File: packages/website/src/theme/MDXComponents/RuleAttributes.tsx [hunk 6]
```diff
@@ -1,4 +1,3 @@
-/* eslint-disable @typescript-eslint/no-unsafe-assignment */
```
Comment (JoshuaKGoldberg) [inline packages/website/src/theme/MDXComponents/RuleAttributes.tsx:1]: [Non-Actionable] Heh, I wonder how much this will come into conflict with #9339...

File: packages/website/src/hooks/useClipboard.ts [hunk 7]
```diff
@@ -7,12 +7,13 @@ export type useClipboardResult = [copied: boolean, copy: () => void];
...
-      setCopied(true);
-    });
-  }, [setCopied, code]);
+  const copy = useCallback(
+    () =>
+      void navigator.clipboard.writeText(code()).then(() => {
+        setCopied(true);
+      }),
+    [setCopied, code],
+  );
```
Comment (JoshuaKGoldberg) [inline packages/website/src/hooks/useClipboard.ts:16]: [Question] Why this change? Seems unrelated & something to be reverted to me?

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
I think we'll need to keep ESLint 8 testing in there for a while - and there is some noise making it harder to review. But encouraging as a start! 🌞

File: packages/website/src/hooks/useClipboard.ts (hunk 7 above)
Comment (abrahamguo) [reply packages/website/src/hooks/useClipboard.ts:16]: Nope — with ESLint 9, it now reports unused `eslint-disable` comments, like this one 🎉

File: packages/website-eslint/build.ts (hunk 4 above)
Comment (abrahamguo) [reply packages/website-eslint/build.ts:34]: sorry about that! prettier is annoyingly un-opinionated about objects on one line vs multiple lines and was conflicting with my editor

File: eslint.config.mjs (hunk 3 above)
Comment (abrahamguo) [reply eslint.config.mjs:522]: @bradzacher that's actually not true.

This can be removed _if_ you are OK with me editing the section at the top of this file labeled `registering all the plugins up-front`, and removing any plugin for which we use a shared config (for example, `reactPlugin` and `reactHooksPlugin`).
Shared configs automatically register their corresponding plugin, so it is unnecessary to manually register the plugin.

The flip side, of course, would be if you felt that it's important to have a complete up-front list of all plugins used in any part of the config.

If we choose to keep the complete up-front list of plugins, then the reason why we have to use the compat utilities twice, is because the plugins are defined twice (again, redundantly so), and we need to make sure that in _both_ places that each plugin is registered, it is the fixed-up version.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (abrahamguo) [reply .github/workflows/ci.yml:137]: We still do have [this integration test for ESLint 8](https://github.com/typescript-eslint/typescript-eslint/tree/main/packages/integration-tests/fixtures/eslint-v8).

@bradzacher's [thoughts from above](https://github.com/typescript-eslint/typescript-eslint/pull/9119#issuecomment-2124749156)

and looking at the history of this file, it seems like we have not had similar tests in the past for other versions of ESLint

Comment (bradzacher) [reply .github/workflows/ci.yml:137]: I don't think we need to worry too much.
There aren't any new APIs we need to worry about for plugins at least.
Given we have extension rules we need to be careful testing across majors. I'm surprised none of our tests broke!

When we had v7+v8 support we only relied on the integration test and not a full CI.
I think this might just be good enough.

File: eslint.config.mjs (hunk 2 above)
Comment (JoshuaKGoldberg) [reply eslint.config.mjs:39]: I'd be in support of this as a separate followup yeah.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (JoshuaKGoldberg) [reply .github/workflows/ci.yml:137]: Ah, great, this is enough for me. Thanks! 🎉

**Review by JoshuaKGoldberg — APPROVED** [review]
Whoohoo! Thanks!! 🚀

File: packages/eslint-plugin/rules.d.ts [hunk 8]
```diff
@@ -75,5 +75,4 @@ export type TypeScriptESLintRules = Record<
 >;
 
 declare const rules: TypeScriptESLintRules;
-// eslint-disable-next-line import/no-default-export
```
Comment (abrahamguo) [inline packages/eslint-plugin/rules.d.ts:78]: this is an unnecessary suppression reported by ESLint 9. It is unnecessary because this rule is already disabled on all `d.ts` files in our eslint config file

**Comment by abrahamguo** [comment]
ESLint 9 displays unused lint suppressions as _warnings_, not errors, so there were 3 unused lint suppressions that were displayed in the pipeline but did not block it. I've removed all of those, and provided an explanation for each one.

File: eslint.config.mjs
Comment (auvred) [inline eslint.config.mjs:1]: > ESLint 9 displays unused lint suppressions as _warnings_, not errors, so there were 3 unused lint suppressions that were displayed in the pipeline but did not block it. I've removed all of those, and provided an explanation for each one.

Hmmm, turns out our `eslint.config.mjs` doesn't have  `linterOptions: { reportUnusedDisableDirectives: 'error' }`. Maybe this is a good chance to turn it on?

---

This seems a bit suspicious to me though: this option has not been enabled for this repository until now. Maybe there are some considerations in favor of leaving this option disabled?

Comment (abrahamguo) [reply eslint.config.mjs:1]: Just want to clarify that it was the upgrade from ESLint 8 to ESLint 9 that caused it to begin reporting unnecessary lint suppressions.
I feel like it makes sense to keep it at the default value, or change the level from `warning` to `error`, but I'm open to further discussion.

Comment (JoshuaKGoldberg) [reply eslint.config.mjs:1]: Yeah I'd +1 switching to `error` as a followup. Great find!

**Review by auvred — APPROVED** [review]
:rocket:
